driver-core/src/main/com/mongodb/internal/connection/ConcurrentPool.java

```diff
@@ -192,23 +179,21 @@ T getImmediateUnfair() {
     }
 
     public void prune() {
-        for (RemovalReportingIterator<T> iter = available.iterator(); iter.hasNext();) {
```

Comment:
This is the only functional behavior change.  It reverts to the previous behavior of this method.  I think it's still correct according to the requirements of the connection pool specification, but want to confirm that with the team.

Reply:
The spec says [_"Remove and close perished available Connections"_](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#:~:text=Remove%20and%20close%20perished%20available%20Connections.), and nothing else about this. So the change must be fine.

Reply:
Thinking more about this, I'm uncomfortable that this violates the spirit of the spec, which is strangely silent on whether the pool is LIFO or FIFO and how maxIdleTime is actually supposed to work in practice.  And even if the spec provides this latitude, it's still a significant behavior change for the Java driver.

To keep roughly the same behavior, and still use the standard `ConcurrentLinkedDeque` class, what about something like this:

```
        for (T cur : available) {
            if (itemFactory.shouldPrune(cur) && (available.remove(cur)) {
                close(cur);
            }
        }
```

Each remove becomes linear instead of constant, but on the other hand pruning should be rare in practice.